book/bayesian_optimization.qmd

```diff
@@ -0,0 +1,844 @@
+
+```{r}
+acq_function = acqf("ei", surrogate = surrogate)
+```
+
+We now want to use the Expected Improvement to choose the next candidate for evaluation:
+
+```{r}
+acq_function$update()
+xydt[, "ei" := acq_function$eval_dt(xydt[, surrogate$x_cols, with = FALSE])]
+```
+
+@fig-bayesian-optimization-ei shows that the Expected Improvement is high in regions where the mean prediction of the Gaussian Process is low but the variance prediction suggests uncertainty:
+
+```{r}
+#| echo: true
+#| label: fig-bayesian-optimization-ei
+#| fig-cap: Expected Improvement based on the mean and variance prediction of the Gaussian Process surrogate model.
+#| fig-alt: FIXME
+ggplot(aes(x = x, y = ei), data = xydt) +
+  geom_line() +
+  labs(y = "Expected Improvement") +
+  theme_minimal()
+```
+
+### Acquisition Function Optimizer
+
+To find the most promising next candidate for evaluation, the acquisition function itself must be optimized.
+Internally, an `r ref("OptimInstance")` is constructed using the acquisition function as an `r ref("bbotk::Objective", "Objective")`.
+
+An acquisition function optimizer is then used to solve this optimization problem.
+Technically, this optimizer is a member of the `r ref("mlr3mbo::AcqOptimizer", "AcqOptimizer")` `R6` class.
+
+Construction requires specifying an `r ref("bbotk::Optimizer", "Optimizer")` as well as a `r ref( "bbotk::Terminator", "Terminator")`:
+
+```{r}
+acq_optimizer = acqo(opt("nloptr", algorithm = "NLOPT_GN_ORIG_DIRECT"),
+  terminator = trm("stagnation", iters = 100, threshold = 1e-5))
+```
+
+When working interactively, the `acq_function` must be specified as well:
+
+```{r}
+acq_optimizer = acqo(opt("nloptr", algorithm = "NLOPT_GN_ORIG_DIRECT"),
+  terminator = trm("stagnation", iters = 100, threshold = 1e-5),
+  acq_function = acq_function)
+```
+
+In this example we use the DIRECT algorithm provided by the `nloptr` package to optimize the Expected Improvement.
+We will terminate the acquisition function optimization if we no longer improve at least `1e-5` over a history of `100` iterations.
+
+```{r}
+candidate = acq_optimizer$optimize()
+candidate
+```
+
+@fig-bayesian-optimization-ei-argmax visualizes the location of the next candidate found by maximizing the Expected Improvement.
+
+```{r}
+#| echo: true
+#| label: fig-bayesian-optimization-ei-argmax
+#| fig-cap: Next candidate point (darkred) obtained by maximizing the Expected Improvement based on the mean and variance prediction of the Gaussian Process surrogate model.
+#| fig-alt: FIXME
+ggplot(aes(x = x, y = ei), data = xydt) +
+  geom_line() +
+  geom_point(aes(x = x, y = acq_ei), colour = "darkred", size = 2, data = candidate) +
+  labs(y = "Expected Improvement") +
+  theme_minimal()
+```
+
+We proceed to evaluate the candidate and would then continue with the next iteration of the loop.
+
+```{r}
+instance$eval_batch(candidate)
+```
+
+## Bayesian Optimization for Black Box Optimization {#sec-bayesian-black-box-optimization}
```

Comment:
I think the problem with this chapter is that it takes quite a long time to get to, IMO, the most important part, which is this section. Here we finally see why we should care about everything above. I think it could benefit from having an example like this very early on to hook users